### PR TITLE
Keep memory a MetadataReader created from provider/PEReader accesses alive until the provider/PEReader is released

### PR DESCRIPTION
**Pull request by tmat**
Fix https://github.com/dotnet/corefx/issues/25946.

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs [hunk 1]
```diff
@@ -266,7 +266,7 @@ public unsafe MetadataReader GetMetadataReader(MetadataReaderOptions options = M
                 }
 
                 AbstractMemoryBlock metadata = GetMetadataBlock();
-                var newReader = new MetadataReader(metadata.Pointer, metadata.Size, options, utf8Decoder);
+                var newReader = new MetadataReader(metadata.Pointer, metadata.Size, options, utf8Decoder, memoryOwner: this);
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: Should PEReader.GetMetadataReader do the same?

Comment (tmat) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: Yup. Good catch.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs [hunk 2]
```diff
@@ -85,7 +85,7 @@ public static unsafe MetadataReader GetMetadataReader(this PEReader peReader, Me
             }
 
             var metadata = peReader.GetMetadata();
-            return new MetadataReader(metadata.Pointer, metadata.Length, options, utf8Decoder);
+            return new MetadataReader(metadata.Pointer, metadata.Length, options, utf8Decoder, memoryOwner: peReader);
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs]: This remark is now out-of-date. Ditto for MetadataReaderProvider?

         /// The caller must keep the <see cref="PEReader"/> alive and undisposed throughout the lifetime of the metadata reader

**Comment by nguerrera** [comment]
Commit description and PR title seem backwards: it is the provider/pe reader that is kept alive, not the metadata reader. Right?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs (hunk 2 above)
Comment (nguerrera) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs]: "undisposed" is still true, but can now leave it to finalizer safely. Right?

**Comment by tmat** [comment]
> Commit description and PR title seem backwards: it is the provider/pe reader that is kept alive, not the metadata reader. Right? 

They keep alive each other :)

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs (hunk 2 above)
Comment (tmat) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/PEReaderExtensions.cs]: Yes. Would still be flagged by IDisposable FxCop rule. I'll remove the "alive"

**Comment by tmat** [comment]
test failure: https://github.com/dotnet/corefx/issues/27390

**Comment by tmat** [comment]
test Windows x86 Release Build  please

**Comment by tmat** [comment]
test NETFX x86 Release Build please

**Comment by tmat** [comment]
test NETFX x86 Release Build  please

**Pull request merged**